File: 719.找出第k小的距离对(TODO)/smallestDistancePair.go
package problem0719

// 维护一个大顶堆, 和373做法完全一致，只是这道题的数据规模来说会超时
func smallestDistancePair(nums []int, k int) int {
	heap := make([]int, 0)
	for i := 0; i < len(nums); i++ {
		for j := i + 1; j < len(nums); j++ {
			val := abs(nums[i] - nums[j])
			if len(heap) < k {
				push(&heap, val)
			} else if less(heap[0], val) {
				pop(&heap)
				push(&heap, val)
			}
		}
	}
	return heap[len(heap)-k]
}

func push(arr *[]int, val int) {
	*arr = append(*arr, val)
	up(*arr, len(*arr)-1)
}

func pop(arr *[]int) {
	lastIndex := len(*arr) - 1
	swap(*arr, 0, lastIndex)
	down(*arr, 0, lastIndex)
	*arr = (*arr)[:lastIndex]
}

func up(arr []int, i int) {
	pIndex := (i - 1) / 2
	if less(arr[i], arr[pIndex]) {
		swap(arr, i, pIndex)
		up(arr, pIndex)
	}
}

func down(arr []int, i int, length int) {
	leftIndex := i*2 + 1
	if leftIndex >= length {
		return
	}
	if leftIndex+1 < length && less(arr[leftIndex+1], arr[leftIndex]) {
		leftIndex++
	}
	if less(arr[leftIndex], arr[i]) {
		swap(arr, i, leftIndex)
		down(arr, leftIndex, length)
	}
}

func swap(arr []int, i, j int) {
	arr[i], arr[j] = arr[j], arr[i]
}

func less(a, b int) bool {
	return a > b
}

func abs(num int) int {
	if num < 0 {
		return -num
	}
	return num
}

File: 719.找出第k小的距离对/smallestDistancePair.js
function smallestDistancePair(nums, k) {
  // 维护一个大小为k的最大堆
  const heap = new Heap([], (a, b) => a > b)
  const existMap = {}
  for (let i = 0; i < nums.length; i++) {
    for (let j = i; j < nums.length; j++) {
      const diff = Math.abs(nums[i] - nums[j])
      if (diff in existMap) {
        continue
      }
      if (heap.length < k) {
        heap.push(diff)
      } else {
        if (diff < heap.peak()) {
          heap.pop()
          heap.push(diff)
        }
      }
    }
  }
  return heap.peak()
}

// 堆
class Heap {
  constructor(data = [], less) {
    this.data = data
    this.less = less || ((a, b) => a < b)

    if (this.length) {
      for (let p = (this.length - 2) >> 1; p >= 0; p--) {
        this._down(p)
      }
    }
  }
  get length() {
    return this.data.length
  }
  peak() {
    return this.data[0]
  }
  push(val) {
    this.data.push(val)
    this._up(this.length - 1)
  }
  pop() {
    if (!this.length) {
      return
    }
    this._swap(0, this.length - 1)
    const popItem = this.data.pop()
    this._down(0)
    return popItem
  }
  _swap(i, j) {
    [this.data[i], this.data[j]] = [this.data[j], this.data[i]]
  }
  _up(i) {
    if (i <= 0) return
    const pIndex = (i - 1) >> 1
    if (this.less(this.data[i], this.data[pIndex])) {
      this._swap(i, pIndex)
      this._up(pIndex)
    }
  }
  _down(i) {
    let leftIndex = i * 2 + 1
    if (leftIndex >= this.length) {
      return
    }
    if (leftIndex + 1 < this.length && this.less(this.data[leftIndex+1], this.data[leftIndex])) {
      leftIndex++
    }
    if (this.less(this.data[leftIndex], this.data[i])) {
      this._swap(leftIndex, i)
      this._down(leftIndex)
    }
  }
}
